code/mob/living/carbon/human.dm

```diff
@@ -1139,13 +1139,29 @@
 					src.set_a_intent(INTENT_DISARM)
 				return
 		else
-			if (src.client.check_key(KEY_THROW) && src.a_intent == "help" && isliving(target) && BOUNDS_DIST(src, target) <= 0)
+			if (src.client.check_key(KEY_THROW) && src.equipped() && src.a_intent == "help" && isliving(target) && BOUNDS_DIST(src, target) <= 0)
 				var/obj/item/thing = src.equipped() || src.l_hand || src.r_hand
 				if (thing)
 					usr = src
 					var/mob/living/living_target = target
 					living_target.give_item()
 					return
+			else if (src.client.check_key(KEY_THROW) && !src.equipped() && BOUNDS_DIST(src, target) <= 0)
+				var/turf/T = get_turf(target)
+				var/list/items = list()
+				var/obj/item/I
+				var/highest_weight = 0
+				for (var/i = 1 to min(10, length(T.contents)))
```

Comment:
```suggestion
				for (var/obj/item/I in T.contents)
```
Doing it this way lets you take advantage of Byond's basic type filtering optimizations and should easily be fast enough to just iterate the entire turf contents. Bear in mind that pathfinding code is doing this for every turf it checks, so doing it maximum once every few seconds on a user action is absolutely fine.

For reference, with a limit of 10 I was finding the weight priority system was breaking on any table with a lot of things on it, partly because it was also iterating lighting overlays, underfloor cables etc.